django/db/models/base.py

```diff
@@ -454,10 +465,12 @@ def __init__(self, *args, **kwargs):
 
     @classmethod
     def from_db(cls, db, field_names, values):
-        if cls._deferred:
-            new = cls(**dict(zip(field_names, values)))
-        else:
-            new = cls(*values)
+        # TODO - make QuerySet iterator fill the missing fields.
```

Comment:
The idea is that we could always have all field_names and values present for the from_db() call. The QuerySet iterator is using some extra effort to supply just the field_names and values which are non-deferred. Instead we could supply all field_names and have DEFERRED objs in the values for those fields which are deferred. This might make the logic in queryset iteration a bit simpler, though there isn't a guarantee of that.

The real TODO is to check if this is actually a good change or not.

Sorry I haven't been able to work on this issue, I have been too busy lately.